libraries/Microsoft.Bot.Builder.LanguageGeneration/LGFile.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            CheckErrors(AllDiagnostics);
+            if (!(scope is IMemory memory))
+            {
+                memory = SimpleObjectMemory.Wrap(scope);
+            }
+
+            var evaluator = new Evaluator(AllTemplates.ToList(), this.expressionEngine);
+            return evaluator.EvaluateTemplate(templateName, new CustomizedMemory(memory));
+        }
+
+        /// <summary>
+        /// Expand a template with given name and scope.
+        /// Return all possible responses instead of random one.
+        /// </summary>
+        /// <param name="templateName">Template name to be evaluated.</param>
+        /// <param name="scope">The state visible in the evaluation.</param>
+        /// <returns>Expand result.</returns>
+        public IList<string> ExpandTemplate(string templateName, object scope = null)
+        {
+            CheckErrors(AllDiagnostics);
+            var expander = new Expander(AllTemplates.ToList(), this.expressionEngine);
+            return expander.EvaluateTemplate(templateName, new CustomizedMemory(scope));
+        }
+
+        /// <summary>
+        /// (experimental)
+        /// Analyzer a template to get the static analyzer results including variables and template references.
+        /// </summary>
+        /// <param name="templateName">Template name to be evaluated.</param>
+        /// <returns>analyzer result.</returns>
+        public AnalyzerResult AnalyzeTemplate(string templateName)
+        {
+            CheckErrors(AllDiagnostics);
+            var analyzer = new Analyzer(AllTemplates.ToList(), this.expressionEngine);
+            return analyzer.AnalyzeTemplate(templateName);
+        }
+
+        /// <summary>
+        /// update an exist template.
+        /// </summary>
+        /// <param name="templateName">origin template name. the only id of a template.</param>
+        /// <param name="newTemplateName">new template Name.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new template body.</param>
+        /// <returns>new LG resource.</returns>
+        public LGFile UpdateTemplate(string templateName, string newTemplateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+
+            var templateNameLine = BuildTemplateNameLine(newTemplateName, parameters);
+            var newTemplateBody = ConvertTemplateBody(templateBody);
+            var content = $"{templateNameLine}\r\n{newTemplateBody}\r\n";
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var newContent = ReplaceRangeContent(Content, startLine, stopLine, content);
+            return LGParser.ParseText(newContent, Id, ImportResolver);
```

Comment:
Should we return a new LGFile or update this LGFile in place?

Reply:
Maybe we can get feedback from @zhixzhan from a caller side, see which pattern is more convenient?